### PR TITLE
8340214: C2 compilation asserts with "no node with a side effect" in PhaseIdealLoop::try_sink_out_of_loop

### PR DESCRIPTION
**Pull request by rwestrel**
The patch includes 2 test cases for this: test1() causes the assert
failure in the bug description, test2() causes an incorrect execution
where a load floats above a store that it should be dependent on.

In the test cases, `field` is accessed on object `a` of type `A`. When
the field is accessed, the type that c2 has for `a` is `A` with
interface `I`. The holder of the field is class `A` which implements
no interface. The reason the type of `a` and the type of the holder
are slightly different is because `a` is the result of a merge of
objects of subclasses `B` and `C` which implements `I`.

The root cause of the bug is that `Compile::flatten_alias_type()`
doesn't change `A` + interface `I` into `A`, the actual holder of the
field. So `field` in `A` + interface `I` and `field` in `A` get
different slices which is wrong. At parse time, the logic that creates
the `Store` node uses:

```
C->alias_type(field)->adr_type()
```

to compute the slice which is the slice for `field` in `A`. So the
slice used at parse time is the right one but during igvn, when the
slice is computed from the input address, a different slice (the one
for `A` + interface `I`) is used. That causes load/store nodes when
they are processed by igvn to use the wrong memory state.

In `Compile::flatten_alias_type()`:

```
if (!ik->equals(canonical_holder) || tj->offset() != offset) {
  if( is_known_inst ) {
    tj = to = TypeInstPtr::make(to->ptr(), canonical_holder, true, nullptr, offset, to->instance_id());
  } else {
    tj = to = TypeInstPtr::make(to->ptr(), canonical_holder, false, nullptr, offset);
  }
}
```

only flattens the type if it's not the canonical holder but it should
test that the type doesn't implement interfaces that the canonical
holder doesn't. To keep the logic simple, the fix I propose creates a
new type whenever there's a chance that a type implements extra
interfaces (the type is not exact).

I also added asserts in `GraphKit::make_load()` and
`GraphKit::store_to_memory()` to make sure the slice that is passed
and the address type agree. Those asserts fire with the new test
cases. When running testing, I found that they also catch a few cases
in `library_call.cpp` where an incorrect slice is passed.

As further clean up, maybe we want to drop the slice argument to
`GraphKit::make_load()` and `GraphKit::store_to_memory()` (and to
their callers) given it's redundant with the address type and error
prone.

/cc hotspot-compiler

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340214](https://bugs.openjdk.org/browse/JDK-8340214): C2 compilation asserts with "no node with a side effect" in PhaseIdealLoop::try_sink_out_of_loop (**Bug** - P2)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21303/head:pull/21303` \
`$ git checkout pull/21303`

Update a local copy of the PR: \
`$ git checkout pull/21303` \
`$ git pull https://git.openjdk.org/jdk.git pull/21303/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21303`

View PR using the GUI difftool: \
`$ git pr show -t 21303`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21303.diff">https://git.openjdk.org/jdk/pull/21303.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21303#issuecomment-2388412676)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/types/TestBadMemSliceWithInterfaces.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            if (res != 42) {
+                throw new RuntimeException("incorrect result " + res);
+            }
+        }
+    }
+
+    private static void test1(B b, C c, boolean flag) {
+        A a;
+        if (flag) {
+           a = b;
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/types/TestBadMemSliceWithInterfaces.java:56]: Indentation is off:
```suggestion
            a = b;
```

File: src/hotspot/share/opto/compile.cpp [hunk 2]
```diff
@@ -1465,12 +1465,15 @@ const TypePtr *Compile::flatten_alias_type( const TypePtr *tj ) const {
     } else {
       ciInstanceKlass *canonical_holder = ik->get_canonical_holder(offset);
       assert(offset < canonical_holder->layout_helper_size_in_bytes(), "");
-      if (!ik->equals(canonical_holder) || tj->offset() != offset) {
-        if( is_known_inst ) {
-          tj = to = TypeInstPtr::make(to->ptr(), canonical_holder, true, nullptr, offset, to->instance_id());
-        } else {
-          tj = to = TypeInstPtr::make(to->ptr(), canonical_holder, false, nullptr, offset);
-        }
+      assert(tj->offset() == offset, "not change to offset expected");
```
Comment (chhagedorn) [inline src/hotspot/share/opto/compile.cpp:1468]:
```suggestion
      assert(tj->offset() == offset, "no change to offset expected");
```

File: src/hotspot/share/opto/compile.cpp [hunk 3]
```diff
@@ -1465,12 +1465,15 @@ const TypePtr *Compile::flatten_alias_type( const TypePtr *tj ) const {
...
+      bool xk = to->klass_is_exact();
+      int instance_id = to->instance_id();
+
+      if (xk && ik->equals(canonical_holder)) {
+        assert(tj == TypeInstPtr::make(to->ptr(), canonical_holder, is_known_inst, nullptr, offset, instance_id), "exact type should be canonical type");
+      } else {
+        assert(xk || !is_known_inst, "Known instance should be exact type");
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.cpp:1475]: Maybe add a comment here and explain the two cases when we create a new type.

**Review by chhagedorn — APPROVED** [review]
Looks reasonable. 

This assert has proven to be quite valuable to find problems in the memory graph that we would otherwise miss. It was also one of the few assert that triggered when having a corrupted graph due to missing Assertion Predicates. I'm wondering if we need more such memory graph checks in general. Anyway, that's just a thought for some future RFE.

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me. I submitted testing and will report back once it passed. 

> As further clean up, maybe we want to drop the slice argument to GraphKit::make_load() and GraphKit::store_to_memory() (and to their callers) given it's redundant with the address type and error prone.

Yes, let's do that. Please file a starter RFE.

**Comment by rwestrel** [comment]
> > As further clean up, maybe we want to drop the slice argument to GraphKit::make_load() and GraphKit::store_to_memory() (and to their callers) given it's redundant with the address type and error prone.
> 
> Yes, let's do that. Please file a starter RFE.

https://bugs.openjdk.org/browse/JDK-8341411

File: src/hotspot/share/opto/compile.cpp (hunk 3 above)
Comment (rwestrel) [reply src/hotspot/share/opto/compile.cpp:1478]: Done in new commit.

Comment (TobiHartmann) [reply src/hotspot/share/opto/compile.cpp:1478]: Thanks!

**Review by TobiHartmann — APPROVED** [review]

File: test/hotspot/jtreg/compiler/types/TestBadMemSliceWithInterfaces.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        for (int i = 0; i < 1000; i++) {
+            a.field = 42;
+        }
+    }
+
+    private static int test2(B b, C c, boolean flag) {
+        A a;
+        if (flag) {
+           a = b;
```
Comment (chhagedorn) [inline test/hotspot/jtreg/compiler/types/TestBadMemSliceWithInterfaces.java:68]:
```suggestion
            a = b;
```

**Review by chhagedorn — APPROVED** [review]
Still good, one more minor thing.

**Review by chhagedorn — APPROVED** [review]

**Review by TobiHartmann — APPROVED** [review]

**Comment by rwestrel** [comment]
@TobiHartmann Do you have an update on testing?

**Comment by TobiHartmann** [comment]
Sorry, that slipped through. Testing looked good. Let me re-run some quick testing with the latest updates.

**Comment by TobiHartmann** [comment]
All testing passed.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews.

**Pull request closed without merging**